File: Week 7  Clock.py/Clock.py
#an example of object orientated programmining 

#NB : difference between OOP and PP - the data is separate from the functions, which manipulate the data

# in OOP , the class contains both the data and algorithms
# method double underlined - reserved method name in python and important function called constructor . Many clocks in a house , but use the same blue
# print. same algorithms but u can use different data

import time
# has control over its own data
class Clock:
    def __init__ (self,hours,mins,secs):
        self.hours = hours
        self.mins = mins
        self.secs = secs
        self.validate()


# check what the person enters , if its not aligned with a clock. it reverts to 0 
    def validate (self):
        if (self.hours > 12 or self.mins >60 or self.secs >60):
            self.hours = 0
            self.mins = 0
            self.secs = 0


    # tick method for moving the clock along , by 1 sec , if it is increase , the hrs and mins increases
    def tick (self):
        self.secs +=1

        if  (self.secs >59):
            self.mins += 1
            self.secs = 0

        if  (self.mins >59):
            self.hours += 1
            self.mins = 0
        
        if  (self.hours > 12):
            self.hours = 1
            self.mins =0
            self.secs = 0


      # this special method returns a instance of the class    , state of the class. returns a string based 
      # self means hours
      # init methods. hours is send into method.   
    def __rep__(self):
        return f"{self.hours}:{self.mins}:{self.secs}"
        
if __name__=='__main__':
    c = Clock(10, 20, 50)
    c2 = Clock(11,23,55);

    while(True):
        c.tick()
        print(c)
        time.sleep(1)

File: Week 7  Clock.py/Clock.c
//here under procedural, the data is in the struct and the methods exist outside the struct. 
//separate
//OOP the data and methods exist together in a class - uses encapulation and inheritances

// the data from the struct has to be passed into the tick method;

#include<stdio.h>
#include<windows.h>

struct Clock {
    int hours;
    int mins;
    int secs;
};

void tick(struct Clock *c)
{
    c -> secs += 1;

    if (c->secs >59)
    {
        c->mins += 1;
        c->secs =0;
    }

    if (c->mins >59)
    {
        c->hours+= 1;
        c->mins =0;
    }


    if (c->hours >12)
    {
        c->hours= 1;
        c->mins =0;
        c->secs =0;
    }
}

// struct clock here is passed as a clock variable/values. No change to data required 

void printClock (struct Clock c)
{
    printf ("\n Clock");
    printf("\n%02d:%02d:%02d:",c.hours,c.mins,c.secs);
}

// struct Clock is passed as a pointer, as data may need to be changed , called passed by reference. 
// 
void validate (struct Clock *c)
{ 
    if (c-> hours >12 || c->mins>60 || c->secs>60)
    {
        c->hours = 0;
        c->mins = 0;
        c->secs = 0;
    }
}

// stript is the same , a new struct is created ( data container), validation happens outside. 
int main()

{
    struct Clock c ={10, 20, 30};
    validate(&c);
// infinite loop : 1000 miliseconds ( 1sec)
    while (1)
    {
        tick (&c);
        printClock(c);
        Sleep(1000);
    }
    return 0;
}







